cli/internal/helm/actionfactory.go

```diff
@@ -0,0 +1,186 @@
+/*
+Copyright (c) Edgeless Systems GmbH
+
+SPDX-License-Identifier: AGPL-3.0-only
+*/
+
+package helm
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/edgelesssys/constellation/v2/internal/compatibility"
+	"github.com/edgelesssys/constellation/v2/internal/constants"
+	"github.com/edgelesssys/constellation/v2/internal/semver"
+	"helm.sh/helm/v3/pkg/action"
+	"helm.sh/helm/v3/pkg/chart"
+)
+
+// ErrConfirmationMissing signals that an action requires user confirmation.
+var ErrConfirmationMissing = errors.New("action requires user confirmation")
+
+var errReleaseNotFound = errors.New("release not found")
+
+type actionFactory struct {
+	versionLister releaseVersionLister
+	cfg           *action.Configuration
+	kubeClient    crdClient
+	cliVersion    semver.Semver
+	log           debugLog
+}
+
+type crdClient interface {
+	ApplyCRD(ctx context.Context, rawCRD []byte) error
+}
+
+// newActionFactory creates a new action factory for managing helm releases.
+func newActionFactory(kubeClient crdClient, lister releaseVersionLister, actionConfig *action.Configuration, cliVersion semver.Semver, log debugLog) *actionFactory {
+	return &actionFactory{
+		cliVersion:    cliVersion,
+		versionLister: lister,
+		cfg:           actionConfig,
+		kubeClient:    kubeClient,
+		log:           log,
+	}
+}
+
+// GetActions returns a list of actions to apply the given releases.
+func (a actionFactory) GetActions(releases []Release, force, allowDestructive bool) (actions []applyAction, includesUpgrade bool, err error) {
+	upgradeErrs := []error{}
+	for _, release := range releases {
+		err := a.appendNewAction(release, force, allowDestructive, &actions)
+		var invalidUpgrade *compatibility.InvalidUpgradeError
+		if errors.As(err, &invalidUpgrade) {
+			upgradeErrs = append(upgradeErrs, err)
+			continue
+		}
+		if err != nil {
+			return actions, includesUpgrade, fmt.Errorf("creating action for %s: %w", release.ReleaseName, err)
+		}
+	}
+	for _, action := range actions {
+		if _, ok := action.(*upgradeAction); ok {
+			includesUpgrade = true
+			break
+		}
+	}
+	return actions, includesUpgrade, errors.Join(upgradeErrs...)
+}
+
+func (a actionFactory) appendNewAction(release Release, force, allowDestructive bool, actions *[]applyAction) error {
```

Comment:
When I first read the function I expected a similar interface to Go's builtin `append` function. I.e. `([]applyAction, error)` as return value. I couldn't find a reason why such an interface wouldn't work. This obviously works as well, no need to change sth now.

Reply:
I think its just safer and more readible since you otherwise need to pass a pointer to the slice

Reply:
This version requires you to pass a pointer. Doing sth like the following doesn't:
```
func appendSth(slice []int) []int {
	return append(slice, 1)
}
```